core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -194,49 +199,84 @@ private static ServiceRequestContext serviceRequestContext() {
      *         {@code false} if the initialization has failed and this context's {@link RequestLog} has been
      *         completed with the cause of the failure.
      */
-    public boolean init(EndpointGroup endpointGroup) {
+    public CompletableFuture<Boolean> init(EndpointGroup endpointGroup) {
         assert endpoint == null : endpoint;
         assert !initialized;
         initialized = true;
 
-        Throwable cause = null;
         try {
             if (endpointGroup instanceof Endpoint) {
-                this.endpointGroup = null;
-                updateEndpoint((Endpoint) endpointGroup);
-                runThreadLocalContextCustomizers();
+                return initEndpoint((Endpoint) endpointGroup);
             } else {
-                this.endpointGroup = endpointGroup;
-                // Note: thread-local customizer must be run before EndpointSelector.select()
-                //       so that the customizer can inject the attributes which may be required
-                //       by the EndpointSelector.
-                runThreadLocalContextCustomizers();
-                updateEndpoint(endpointGroup.select(this));
+                return initEndpointGroup(endpointGroup);
             }
         } catch (Throwable t) {
-            cause = t;
+            acquireEventLoop(endpointGroup);
+            failEarly(t);
+            return UnmodifiableFuture.completedFuture(false);
         }
+    }
 
-        if (eventLoop == null) {
-            final ReleasableHolder<EventLoop> releasableEventLoop =
-                    options().factory().acquireEventLoop(sessionProtocol(), endpointGroup, endpoint);
-            eventLoop = releasableEventLoop.get();
-            log.whenComplete().thenAccept(unused -> releasableEventLoop.release());
+    private UnmodifiableFuture<Boolean> initEndpoint(Endpoint endpoint) {
+        this.endpointGroup = null;
+        updateEndpoint(endpoint);
+        runThreadLocalContextCustomizers();
+        acquireEventLoop(endpoint);
+        return UnmodifiableFuture.completedFuture(true);
```

Comment:
How about making this private static final and reusing it?

Reply:
It's reused inside `UnmodifiableFuture`.

Reply:
Ah I missed that as well. 😅